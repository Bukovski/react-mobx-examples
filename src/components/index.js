export { TodoNew } from './todo-new.component';
export { TodoList } from './todo-list.component';
export { TodoItem } from './todo-item.component';
