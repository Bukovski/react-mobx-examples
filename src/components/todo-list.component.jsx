import React from 'react';
import { TodoItem } from "./index";
import { useStore } from "../helpers/store-context.helper";


export const TodoList = () => {
	const todoList = useStore();
	
	return (
		<div className="todo-list">
			<ul className="open-todos todos align">
				<span>Open Todos</span>
				{
					todoList.openTodos.map(todo =>
						<TodoItem key={ todo.id } todo={ todo } />
					)
				}
			</ul>
			<ul className="finished-todos todos align">
				<span>Finished Todos</span>
				{
					todoList.finishedTodos.map(todo =>
						<TodoItem key={ todo.id } todo={ todo } />
					)
				}
			</ul>
		</div>
	)
};