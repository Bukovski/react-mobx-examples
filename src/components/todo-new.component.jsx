import React, { useState } from 'react';


export const TodoNew = () => {
	const [ todo, setTodo ] = useState('');
	
	const addTodo = () => {
		setTodo('');
	};
	
	return (
		<div className="add">
			<input
				type="text"
				name="add"
				placeholder="Add item..."
				value={ todo }
				onChange={ (e) => setTodo(e.target.value) }
			/>
				<div className="input-buttons" onClick={ addTodo }>
					<button className="add-todo">
						<i className="fas fa-plus add plus-icon"/>
					</button>
				</div>
		</div>
	)
};
