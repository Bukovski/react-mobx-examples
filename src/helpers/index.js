export { StoreContext, StoreProvider, useStore } from './store-context.helper';
export { randomId } from './random-id.helper';
export { onEnterPress } from './use-enter.helper';
