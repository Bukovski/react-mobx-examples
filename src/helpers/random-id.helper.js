export const randomId = () => {
	return '_'
		+ (Date.now().toString(36)
			+ Math.random().toString(36).substr(2, 9))
			.toUpperCase()
};
