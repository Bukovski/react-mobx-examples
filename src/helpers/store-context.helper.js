import { createContext, useContext } from 'react';


export const StoreContext = createContext({});
export const StoreProvider = StoreContext.Provider;

export const useStore = () => useContext(StoreContext);
