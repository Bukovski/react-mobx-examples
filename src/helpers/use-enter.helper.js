export const onEnterPress = (callback) => {
	return (event) => {
		if (event.key === 'Enter') {
			callback();
		}
	}
};
