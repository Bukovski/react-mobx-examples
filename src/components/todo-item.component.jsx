import React, { useState } from 'react';
import { useStore, onEnterPress } from "../helpers";


export const TodoItem = (props) => {
	const { todo } = props;
	
	const todoListStore = useStore();
	
	const [ text, setText ] = useState('');
	const [ edit, setEdit ] = useState(false);
	
	const saveText = () => {
		todo.updateText(text);
		
		setEdit(false);
		setText('');
	};
	
	return (
		<div className="todo-item">
			{
				edit
					? <li className="edit-item">
						<input
							type="text"
							onKeyDown={ onEnterPress(saveText) }
							onChange={ (event) => setText(event.target.value) }
						/>
						<i className="far fa-save save" onClick={ saveText }/>
					</li>
					: <li>
						<input
							type="checkbox"
							id={`todo_${ todo.id }`}
							defaultChecked={ todo.isDone }
							onChange={ todo.toggleIsDone }
						/>
						<label htmlFor={`todo_${ todo.id }`}>
							<span className="check" />
							{ todo.text }
						</label>
						<div className="todo-buttons">
							<i className="far fa-edit edit" onClick={ () => setEdit(true) }/>
							<i className="far fa-trash-alt delete" onClick={ () => todoListStore.removeTodo(todo) }/>
						</div>
					</li>
			}
		</div>
	)
};
