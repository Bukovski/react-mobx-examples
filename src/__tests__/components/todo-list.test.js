import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { StoreProvider } from "../../helpers";
import { TodoList } from "../../components";
import TodoListStore from "../../store/todo-list.store";
import * as IdRandom from "../../helpers/random-id.helper";




describe("TodoList", () => {
	let todoListStore;
	
	beforeAll(() => {
		function* idMaker(){
			yield "1";
			yield "2";
			yield "3";
			yield "4";
			yield "5";
		}
		const idGenerator = idMaker();
		
		jest.mock("../../helpers/random-id.helper")
		IdRandom[ "randomId" ] = () => idGenerator.next().value;
		
		todoListStore = new TodoListStore([
			"React", "Redux",	"MobX"
		]);
	})
	
	afterAll(() => {
		jest.clearAllMocks();
	});
	
	it("render page with provider props", () => {
		const { container } = render(
			<StoreProvider value={ todoListStore }>
				<TodoList/>
			</StoreProvider>
		);
		
		expect(screen.getByText("Redux")).toBeInTheDocument();
		expect(screen.getByText("Redux").closest("ul")).toHaveClass("open-todos");
		
		userEvent.click(screen.getByRole("checkbox", { name: "Redux" }))
		expect(screen.getByText("Redux").closest("ul")).toHaveClass("finished-todos");

		expect(container.firstChild).toMatchSnapshot();
	});
	
	it("edit item from list", () => {
		render(
			<StoreProvider value={ todoListStore }>
				<TodoList/>
			</StoreProvider>
		);
		
		expect(screen.getByText("React")).toBeInTheDocument();

		userEvent.click(screen.getByText("React").closest("li").querySelector('i.edit'));
		
		const textbox = screen.getByRole("textbox");
		
		expect(textbox).toHaveValue("React");
		
		userEvent.type(textbox, " new text");
		expect(textbox).toHaveValue("React new text");
		
		userEvent.click(textbox.closest("li").querySelector('i.save'));
		expect(screen.getByText("React new text")).toBeInTheDocument();
	});
	
	it("remove item from list", () => {
		render(
			<StoreProvider value={ todoListStore }>
				<TodoList/>
			</StoreProvider>
		);
		
		expect(screen.getByText("MobX")).toBeInTheDocument();
		expect(todoListStore.list).toHaveLength(3)
		
		userEvent.click(screen.getByText("MobX").closest("li").querySelector('i.delete'));
		expect(screen.queryByText("MobX")).not.toBeInTheDocument();
		expect(todoListStore.list).toHaveLength(2)
	});
})