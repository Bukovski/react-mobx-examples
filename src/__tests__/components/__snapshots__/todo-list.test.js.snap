// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList render page with provider props 1`] = `
<div
  class="todo-list"
>
  <ul
    class="open-todos todos align"
  >
    <span>
      Open Todos
    </span>
    <div
      class="todo-item"
    >
      <li>
        <input
          id="todo_1"
          type="checkbox"
        />
        <label
          for="todo_1"
        >
          <span
            class="check"
          />
          React
        </label>
        <div
          class="todo-buttons"
        >
          <i
            class="far fa-edit edit"
          />
          <i
            class="far fa-trash-alt delete"
          />
        </div>
      </li>
    </div>
    <div
      class="todo-item"
    >
      <li>
        <input
          id="todo_3"
          type="checkbox"
        />
        <label
          for="todo_3"
        >
          <span
            class="check"
          />
          MobX
        </label>
        <div
          class="todo-buttons"
        >
          <i
            class="far fa-edit edit"
          />
          <i
            class="far fa-trash-alt delete"
          />
        </div>
      </li>
    </div>
  </ul>
  <ul
    class="finished-todos todos align"
  >
    <span>
      Finished Todos
    </span>
    <div
      class="todo-item"
    >
      <li>
        <input
          checked=""
          id="todo_2"
          type="checkbox"
        />
        <label
          for="todo_2"
        >
          <span
            class="check"
          />
          Redux
        </label>
        <div
          class="todo-buttons"
        >
          <i
            class="far fa-edit edit"
          />
          <i
            class="far fa-trash-alt delete"
          />
        </div>
      </li>
    </div>
  </ul>
</div>
`;
